Move props from datastore to order object and implement $save params for DataStore

vedanshujain opened the pull request:
### All Submissions:

-   [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
-   [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
-   [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change?

<!-- Mark completed items with an [x] -->

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

We currently store many props in the metadata, and they are set/get directly on the go by the CPT datastore. However, this is problematic because they are technically props of the order, not the datastore. DataStore does not have any representation in the database. 

This causes a problem which is that we have moved some of these props to the custom table. Given that they are directly being manipulated in the post meta table, we end storing duplicates of them (one in the operation table, and then again in the meta (which may be added by the WC core itself with $order->update_meta calls). 

To solve this, we have moved the properties to the WC_Order object itself. This visibility is public, but since datastore was already exposing these props with public visibility, it's the same. This PR also adds a new internal var in the order object called `legacy_datastore_props`; the goal of this prop is that whenever `add/update/delete` meta is called for any of these internal props, we call the respective getter/setter instead, without throwing a warning.

### How to test the changes in this Pull Request:

Adequate unit tests have been added for setting/updating these props, so no additional testing is necessary. However, you can test that no duplicate data is added for these props.

The bug that these props being in metadata would cause is that we would be adding duplicate metadata even though the prop already exists in the table. In this case, props in the operational table would not be updated, but functionality won't be affected since we would add meta rows, and they will be updated as expected. So no visible functionality change can be observed from the UX.

### Other information:

-   [x] Have you added an explanation of what your changes do and why you'd like us to include them?
-   [x] Have you written new tests for your changes, as applicable?
-   [x] Have you successfully run tests with your changes locally?
-   [x] Have you created a changelog file for each project being changed, ie `pnpm changelog add --filter=<project>`?

<!-- Mark completed items with an [x] -->

### FOR PR REVIEWER ONLY:

-   [x] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

vedanshujain committed:
Change order data store internal key to props for better representation.

vedanshujain committed:
Add changelog

vedanshujain committed:
Merge branch 'trunk' into implement/data_store_saves

vedanshujain committed:
Add legacy datastore prop for WC_Data object to support COT.

vedanshujain committed:
Also support passing in `get_stock_reduced` method.

vedanshujain committed:
Add internal meta keys to order table data store.

vedanshujain committed:
Merge branch 'trunk' into implement/data_store_saves

vedanshujain committed:
Add tests for CPT datastore for orders.

vedanshujain committed:
Change return type to prevent implicit type conversion.

vedanshujain committed:
Use correct conversion.

vedanshujain committed:
Revert "Use correct conversion."

This reverts commit 97ead11f717e42fecff504feaceb99bf567098e2.

vedanshujain committed:
Add conversions to test as needed.

vedanshujain committed:
Add internal keys for order table data store for consistency.

vedanshujain committed:
Apply code standards.

vedanshujain committed:
Merge branch 'trunk' into implement/data_store_saves

jorgeatorres commented on plugins/woocommerce/includes/class-wc-order.php:
```diff
@@ -975,6 +982,61 @@ public function has_shipping_address() {
+	 * @return bool
+	 */
+	public function get_order_stock_reduced( string $context = 'view' ) {
+		return wc_bool_to_string( $this->get_prop( 'order_stock_reduced', $context ) );
```
This should be `wc_string_to_bool()` instead as the expected return value is a bool.

jorgeatorres commented on plugins/woocommerce/includes/class-wc-order.php:
```diff
@@ -975,6 +982,61 @@ public function has_shipping_address() {
+	 * @return bool True if permissions were generated, false otherwise.
+	 */
+	public function get_download_permissions_granted( string $context = 'view' ) {
+		return wc_bool_to_string( $this->get_prop( 'download_permissions_granted', $context ) );
```
`wc_string_to_bool()` should be used here too.

jorgeatorres commented on plugins/woocommerce/includes/class-wc-order.php:
```diff
@@ -975,6 +982,61 @@ public function has_shipping_address() {
+	 * @return bool
+	 */
+	public function get_new_order_email_sent( string $context = 'view' ) {
+		return wc_bool_to_string( $this->get_prop( 'new_order_email_sent', $context ) );
```
`wc_string_to_bool()` should be used here too.

jorgeatorres commented on plugins/woocommerce/includes/class-wc-order.php:
```diff
@@ -975,6 +982,61 @@ public function has_shipping_address() {
+	 * @return bool True if sales were recorded, false otherwise.
+	 */
+	public function get_recorded_sales( string $context = 'view' ) {
+		return wc_bool_to_string( $this->get_prop( 'recorded_sales', $context ) );
```
`wc_string_to_bool()` should be used here too.

jorgeatorres commented on plugins/woocommerce/includes/class-wc-order.php:
```diff
@@ -975,6 +982,61 @@ public function has_shipping_address() {
+	 * @return bool True if coupon counts were updated, false otherwise.
+	 */
+	public function get_recorded_coupon_usage_counts( $context = 'view' ) {
+		return wc_bool_to_string( $this->get_prop( 'recorded_coupon_usage_counts', $context ) );
```
`wc_string_to_bool()` should be used here too.

jorgeatorres commented on plugins/woocommerce/includes/class-wc-order.php:
```diff
@@ -1363,6 +1425,61 @@ public function set_cart_hash( $value ) {
+	 * @return void
+	 */
+	public function set_order_stock_reduced( $value ) {
+		$this->set_prop( 'order_stock_reduced', wc_bool_to_string( $value ) );
```
These applies to all the new properties, but I'm adding the comment here so that we discuss just once:

I think it's a bit confusing that internally we are storing these properties as 'yes' or 'no' values when we really mean true/false (as that's what the getters either in the order or the datastore return).

I understand this was handy when these props were stored in metadata but we're no longer doing that in COT, and it also seems like a datastore-specific thing (whether the datastore uses yes/no or a column or whatever) and not necessarily how the property should live internally in the `WC_Order` object, which to me should be true/false.

I still think we should allow using "yes" or "no" in the setters, though, and specifically the CPT datastore could convert to yes/no when saving (in the `update_post_meta()` method) and convert from yes/no to bool when reading (`read_order_data()`).

I understand this adds a bit of complexity to the CPT implementation for little gain, but it felt more "correct" to me. In any case, this is just my opinion, so I clearly wanted to discuss with you.
Also, definitely not a blocker as everything can work with this as-is (COT automatically does the conversion when formatting values and CPT is already using yes/no) *and* we're talking about how this is internally stored in the order object, not a public interface. In fact, the public interface already converts the internal value to bool.

vedanshujain replied:
I think this problem originated because anyone can do any of `true`, `yes`, `1` (and their opposites) to save data as you point out. This is not a problem in HPOS anymore because the column has bool type, so it's enforced correctly. 

CPT already has the implementation that you suggest, unless I missed something.

vedanshujain replied:
Added storing in order props as bool here https://github.com/woocommerce/woocommerce/pull/34627/commits/c302a85a7ef31e9258e77341629164005d5fe17c

jorgeatorres commented on plugins/woocommerce/src/Internal/DataStores/Orders/OrdersTableDataStore.php:
```diff
@@ -464,8 +525,17 @@ public function get_download_permissions_granted( $order ) {
 	 * @param bool      $save Whether to persist changes to db immediately or not.
 	 */
 	public function set_download_permissions_granted( $order, $set, $save = true ) {
```
(Another discussion that applies to all props)

When we initially added the `$save` arg it was for performance reasons, as we didn't want the datastore level methods to always trigger a save, but now that these props can be set at the order level, wouldn't it be more "compatible" if the COT datastore methods match those in CPT exactly (i.e. get rid of `$save`) and always call `$order->save()`?

We could mark these methods as deprecated in COT and suggest the order-level methods are used instead, but if you do call the datastore-level ones you'll get the exact same behavior as CPT, which seems reasonable to me.
Also, it'd avoid the "no op" scenario you've covered with a `_doing_it_wrong()`.

What do you think?

vedanshujain replied:
Makes sense, not that we are storing data in order prop, they can be used when performance is needed. Removed the param in https://github.com/woocommerce/woocommerce/pull/34627/commits/0cb6cadf0a6054724e02074c552ca1c66bffc6d2

jorgeatorres commented on plugins/woocommerce/src/Internal/DataStores/Orders/OrdersTableDataStore.php:
```diff
@@ -1808,82 +1911,13 @@ public function update_meta( &$object, $meta ) {
 	 *
 	 * @return array List of keys.
 	 */
 	protected function get_internal_data_store_keys() {
```
This method I added while we were finalizing the design, but it's not not a general datastore method but something specific of this class. I believe it should be safe to get rid of it (and the only place where it's used in this class) as ideally there won't be datastore-specific order props (that's exactly what we're fixing here 😅).

vedanshujain replied:
👍 removed in https://github.com/woocommerce/woocommerce/pull/34627/commits/ec0afd608441279ef88e7e2748959791a021521d

jorgeatorres reviewed (changes requested):
Hi @vedanshujain,

Great job on this one 💯. It looks good to me overall, and it's great seeing these "soft" props properly become proper order props (😝).

The only real blocker I saw was that the order-level getter methods are returning "yes" or "no" instead of a boolean, which doesn't match both the docblock and what the datastore-level getter methods do.

I left other comments with a a couple of things to discuss. I'd love to see what you think about those, but I won't hold off on approving if in the end you prefer to leave things as-is.

Please let me know what you think!

vedanshujain committed:
Merge branch 'trunk' into implement/data_store_saves

vedanshujain committed:
Remove $save param for consistency with parent class.

Earlier, we added the $save parameter for performance reason, but now that we are storing the props in the WC_Order object, this isn't needed anymore (as the object can hold the updated props).

vedanshujain committed:
Return bool instead of string as described.

vedanshujain committed:
Fix test by transforming bool values before confirming.

vedanshujain committed:
Fix test by not needing to converting to string, since we now bool value.

vedanshujain committed:
Store internal props as bool instead of string.

vedanshujain committed:
Remove unused method.

jorgeatorres reviewed (changes requested):
Hey @vedanshujain!

Thanks for making the updates. A couple of things caught my eye while testing, but hopefully those should be easily fixable:

1. (CPT-only) For reasons unknown, the `_new_order_email_sent` property is meant to be persisted as the string "true" (not as yes/no nor a boolean). ([See the e-mail class](https://github.com/woocommerce/woocommerce/blob/6e862af9ee4981bd3f2018ad06a63a0c519c1897/plugins/woocommerce/includes/emails/class-wc-email-new-order.php#L113)).

   <img width="414" alt="Screen Shot 2022-09-23 at 11 28 00" src="https://user-images.githubusercontent.com/184724/191990642-9548ceb5-cb2f-44b5-872d-8a19826a5053.png">

   For backwards compatibility, we should probably preserve this behavior in CPT: that is, change `WC_Order_Data_Store_CPT::set_email_sent()` so that it saves the value in the required format _and_ also in `WC_Order_Data_Store_CPT ::update_post_meta()`. Otherwise, because the getter returns a boolean, the metadata that gets saved is `1` (or empty) which is not the same.
2. (CPT & COT) The code in the e-mail class specifically uses `'true' === $email_already_sent` which won't work anymore as the order getter returns a boolean. This would have to be updated too.
3. (CPT-only) While I did advocate for using booleans internally for everything, this also means some conversion is necessary when persisting the order in the CPT datastore. Otherwise, instead of the backwards-compatible "yes" or "no" in the metadata, you get `1` (or empty) when making changes to the props via the order object and not the datastore.
   To reproduce, create an order and compare the results (in the database) of using the datastore vs the order object:
   
   ```php
   $order = wc_get_order( 1234 );
   $order->get_data_store()->set_download_permissions_granted( $order->get_id(), true ); // Using the datastore persists the prop as yes/no.
   ```
   
   vs
   
   ```php
   $order = wc_get_order( 1234 );
   $order->set_download_permissions_granted( true );
   $order->save();
   ```
   
   I think some special handling is needed in [`WC_Order_Data_Store_CPT::update_post_meta()`](https://github.com/woocommerce/woocommerce/blob/ec0afd608441279ef88e7e2748959791a021521d/plugins/woocommerce/includes/data-stores/class-wc-order-data-store-cpt.php#L206) specifically for these props, because the order getter returns a boolean and we need it as yes/no before [the call to `WC_Order_Data_Store_CPT ::update_or_delete_post_meta()`](https://github.com/woocommerce/woocommerce/blob/ec0afd608441279ef88e7e2748959791a021521d/plugins/woocommerce/includes/data-stores/class-wc-order-data-store-cpt.php#L240).

Please let me know what you think!

vedanshujain committed:
Merge branch 'trunk' into implement/data_store_saves

vedanshujain committed:
Only select props for refunds that are actually needed.

vedanshujain committed:
Store bool as yes/no to maintain compatibility with how were storing prev.

vedanshujain committed:
Fix unit test to extpect yes.

vedanshujain commented:
> For reasons unknown, the _new_order_email_sent property is meant to be persisted as the string "true" (not as yes/no nor a boolean)

I see; thanks for catching it. Seems like this field is used more plugins. In this case, we can't consider the field `_new_order_email_sent` as bool, but it should be string (at least as far as code APIs are concerned). 

> While I did advocate for using booleans internally for everything, this also means some conversion is necessary when persisting the order in the CPT datastore. Otherwise, instead of the backwards-compatible "yes" or "no" in the metadata, you get 1 (or empty) when making changes to the props via the order object and not the datastore.

Agreed, added some specific handling for bool keys to store them as yes/no. When the order table becomes the default, hopefully, all this tech debt will go away.

Made both of these changes in https://github.com/woocommerce/woocommerce/pull/34627/commits/7f9cabe638aad308c9a73930ce2c86eff2897e5f. . 

Note that I have added one more bug fix in https://github.com/woocommerce/woocommerce/pull/34627/commits/e2381c9aac7ff0f9beb55f4a1a9d9b6ec48b7354 which came about from changes in this PR. Basically, since refunds also inherits from order, they were trying to set props that didn't existed. I have moved the common methods to abstract class, and removed the props that refunds was not using at all.

jorgeatorres reviewed (changes requested):
Hi @vedanshujain!

Thanks for making the updates. This is definitely looking better. Also, sorry to keep holding this off, but there are still a few tiny issues:

1. For backwards compatibility, when CPT is active, `order_stock_reduced` must be converted to yes/no at save time. Currently, the code is trying to store it as a boolean directly when saving an order after `$order->set_order_stock_reduced()` is called.
   This isn't currently working, despite `stock_reduced` being in the `$internal_data_store_key_getters` array because the property is actually called `order_stock_reduced` in `WC_Order`.
2. `WC_Order::get_new_order_email_sent()` should return a boolean for consistency with the datastore method (and the other methods in general). The e-mail class [is comparing with `'true'`](https://github.com/woocommerce/woocommerce/blob/6e862af9ee4981bd3f2018ad06a63a0c519c1897/plugins/woocommerce/includes/emails/class-wc-email-new-order.php#L106) directly, so that e-mail code needs to be updated anyways.
3. Related to the above, when CPT is active, the `_new_order_email_sent` meta should have a value of `'true'` (that is, the literal "true" string) not yes/no nor a boolean. It's currently being stored as yes/no, but this is a very special case in which we should store it as either `'true'` or `''`.
   
   This doesn't mean we can skip updating the e-mail class mentioned above because the internal order getter would still perform a conversion anyways, but in any case we shouldn't change how the value is stored in the database, in case there are queries relying on that.
   
Please let me know what you think.

vedanshujain committed:
Add tests and set special metadata migrated from data store correctly.

vedanshujain committed:
Add tests for legacy prop get/set in ordertabledatastore.

vedanshujain committed:
Update email class to use data store method instead of get_meta.

vedanshujain committed:
Update exisiting tests to account for changes in legacy prop setting.

vedanshujain commented:
Oh my, agree with all your points. I have added testcases to manage this complexity for both [CPT data store](https://github.com/woocommerce/woocommerce/pull/34627/files#diff-09eea5d10397a66c2d0ef0be7de5e9457e520da176a350f1c859e6e1626f581cR133-R179) and [order table store](https://github.com/woocommerce/woocommerce/pull/34627/files#diff-b1e9beb010bd1b0876aeb1d0ee57cbd0e818998036fffc953b43f2afcd595fc6R1182-R1214). I think this should cover all the cases listed, let me know what you think!

jorgeatorres committed:
Add support for order IDs to prop getters in `OrdersTableDatastore`

jorgeatorres committed:
`OrdersTableDatastore::get_download_permissions_granted()` should no longer rely on metadata

jorgeatorres reviewed (approved):
Hi @vedanshujain!

This LGTM. Thanks for all your work on this one 🥇.

I added a couple of commits (c4255295d36037eda6c4a505db40063cccad2507 and 8ab0278bc749a5acee981e52038f7641076a9d76) that just add a tiny check for the datastore getters in the COT datastore so that it works with order IDs too. This was already in other getters so I just copy/pasted it.

benignant commented on plugins/woocommerce/src/Internal/DataStores/Orders/OrdersTableDataStore.php:
```diff
@@ -568,21 +644,22 @@ private function set_new_order_email_sent( $order, $set, $save = true ) {
 	public function get_stock_reduced( $order ) {
-		$order = is_numeric( $order ) ? wc_get_order( $order ) : $order;
-		return wc_string_to_bool( $order->get_meta( '_order_stock_reduced', true ) );
+		$order = is_int( $order ) ? wc_get_order( $order ) : $order;
```
@vedanshujain This breaks backward compatibility. The implementation of this method for CPT accepts a numeric string. It may be better to use `WC_Order_Factory::get_order_id()` just like `WC_Order_Data_Store_CPT::get_stock_reduced()` does.
